### PR TITLE
Cleanup the API for DateInterval

### PR DESCRIPTION
**Pull request by jonas**
- Makes DateInterval more immutable
 - Only allow creation of DateInterval via DateIntervalType API
 - Use internal builder to ensure truncation of DateInterval.from

### CODE REVIEW

File: core/src/main/scala/com.mediative.amadou/DateInterval.scala [hunk 1]
```diff
@@ -55,13 +53,12 @@ case class DateInterval(
    * }}}
    */
   def to(other: DateInterval): DateInterval =
-    copy(toOpt = Some(other))
-
-  def +(delta: Int): DateInterval = DateInterval(interval) { cal =>
-    cal.setTimeInMillis(from.getTime)
-    cal.add(interval.calendarField, delta * interval.deltaMultiplier)
-  }
+    if (toOpt.forall(_ < other))
```
Comment (yawaramin) [inline core/src/main/scala/com.mediative.amadou/DateInterval.scala:56]: Can also be written `toOpt.forall(other.>=)`.

File: core/src/main/scala/com.mediative.amadou/DateInterval.scala [hunk 2]
```diff
@@ -72,14 +69,15 @@ case class DateInterval(
...
+    interval.defaultFormatter.format(asDate) + toOpt.fold("")(":" + _.toString)
 
   def format(dateFormat: String): String = {
     val formatter = interval.formatter(dateFormat)
-    formatter.format(from) + toOpt.fold("")(":" + _.toString)
+    formatter.format(asDate) + toOpt.fold("")(":" + _.toString)
   }
 
-  def toTimestamp: Timestamp = new Timestamp(from.getTime)
+  def asDate = new Date(from)
```
Comment (yawaramin) [inline core/src/main/scala/com.mediative.amadou/DateInterval.scala:79]: The mutable `Date` is still public, did you change your mind about not exposing it?
^

**Review by yawaramin — APPROVED** [review]
+1 pending a couple of comments.

File: core/src/main/scala/com.mediative.amadou/DateInterval.scala (hunk 2 above)
Comment (jonas) [reply core/src/main/scala/com.mediative.amadou/DateInterval.scala:79]: The mutable date used to be public via `val from = interval.truncate(timestamp)`. However, now we only provide methods to get `Date` and `Timestamp` instances, but they will create a new instance each time.

**Pull request merged**